src/compiler/utilities.ts

```diff
@@ -7119,10 +7119,11 @@ namespace ts {
         }
     }
 
-    export function nodeIsFirstNodeAtOrAfterPosition(sourceFile: SourceFile, node: Node, position: number): boolean {
+    export function nodeIsFirstNodeAtOrAfterPosition(node: Node, position: number): boolean {

```

Comment:
Given my comment for `siblingNodePositionsAreComparable` above, if this function is limited to "adjacent parse tree nodes belonging to the same parent", you can determine if they are truly adjacent (i.e., their order hasn't changed) in two much less expensive ways:

1. Examine the `end` of `previousNode` and the `pos` of `nextNode` and use the context of the parent to determine what, if any, separator token may be present. You can then use `getSourceFileOfNode` (since these nodes are parse tree nodes), `skipTrivia`, and the size of the separator token (if present) to check adjacency.
2. Alternatively, (and possibly even more cheaply) you can use the context of the `parent` node to find which `NodeArray` each node belongs two and use `indexOf` to test for positions. While some of our `Nodes` have multiple `NodeArray` properties, they don't overlap (i.e., for a function-like, `Expression` nodes are probably `arguments`, while `TypeNode` nodes are probably `typeArguments`, etc.).

Reply:
I had thought of something like both of those before. I think option (2) is probably best; it just requires writing a tediously large switch statement on the parent kind. But maybe not too tedious and not too large. For option (1), I don’t think size is sufficient to tell whether something is a separator or not:

```ts
interface I {
  a
  b
  c
}
```

Are `a` and `c` adjacent? There’s only one character of non-trivia between them, but I don’t know whether it’s actually a comma/semicolon or another element without actually looking at the text between them, which seemed more tedious than the former approach. I also do think `skipTrivia` will probably be more expensive than option (1).